### PR TITLE
Ues std::optional (or thrust::optional) for optional parameters & first part of DCSR (DCSC) implementation.

### PR DESCRIPTION
**Pull request by seunghwak**
- Update graph data structure to optionally use DCSR (DCSC) for hypersparse segments
- Use std::optional (or thrust::optional) for optional parameters

Touches many files but a big chunk of the updates are related to std::optional support.
Currently, DCSR (DCSC) is disabled (and primitives are not updated to support the format).
Splitting DCSR (DCSC) support in two PRs to avoid the (already) big PR getting even bigger.

### CODE REVIEW

**Comment by seunghwak** [comment]
rerun tests

**Comment by seunghwak** [comment]
@Iroy30 Could you review the python part of this PR (especially the code passing segment_offsets returned from renumbering to the populate_graph function).
https://github.com/rapidsai/cugraph/pull/1676/commits/f45be2de73a44ed29441459e5a37768883c4cba9 is the most relevant commit.

@kaatish Please review this. You can just skim the updates related to std::optional but you may take a closer look for the DCSR (DCSC) related updates (include/cugraph/experimental/(graph.hpp & graph_view.hpp), src/experimental/(graph.cu & graph_view.cu), src/experimental/renumber_edgelist.cu, ...).

We renumber vertices to ensure that the vertices are sorted by their vertex degree in the descending order within a vertex partition (vertex partition size = # vertices / # GPUs, approximately).

Within a vertex partition, edges of the vertices with their degree equal to or larger than col_comm_size * hypersparse_threshold_ratio are stored in the CSR (or CSC) format.

Edges of the vertices with their degree smaller than col_comm_size * hypersparse_threshold_ratio are stored in the DCSR (or DCSC) format.

Currently, hypersparse_threshold_ratio is set to 0, so this feature is disabled, but it will be enabled once we update primitives to handle the hypersparse segment.

Understanding this is unnecessary to use primitives but will be necessary to update primitives.

**Review by aschaffer — APPROVED** [review]

File: cpp/include/cugraph/experimental/graph.hpp [hunk 1]
```diff
@@ -22,18 +22,21 @@
...
+#include <optional>
 #include <string>
 #include <type_traits>
 #include <vector>
 
 namespace cugraph {
 namespace experimental {
 
+// FIXME: consider using std::optional for optional parameters (i.e. weights & segment_offsets)
+
```
Comment (kaatish) [inline cpp/include/cugraph/experimental/graph.hpp:34]: Is this comment still relevant? I think you changed them to optional already right?

Comment (seunghwak) [reply cpp/include/cugraph/experimental/graph.hpp:34]: Good catch!!! No, I will delete.

**Review by ChuckHastings — APPROVED** [review]
Phew!

One little spelling error I noticed.

**Review by kaatish — APPROVED** [review]

**Review by BradReesWork — APPROVED** [review]

**Comment by seunghwak** [comment]
rerun tests

**Comment by seunghwak** [comment]
rerun tests

**Review by Iroy30 — APPROVED** [review]

**Comment by BradReesWork** [comment]
@gpucibot merge

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by seunghwak** [comment]
rerun tests

**Pull request merged**